### PR TITLE
[Filebeat] backwards compatibility for set processor

### PR DESCRIPTION
**Pull request by leehinman**
## What does this PR do?

When loading a pipeline this change checks the elasticsearch version
and if the version is less than 7.9.0 it will replace the
"ignore_empty_value" option with an equivalent if statement on the set
processor.

## Why is it important?

This allows filebeat > 7.9.0 to be used with older versions of
elasticsearch.  Without it the pipelines fail to load because the
option isn't supported.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally
run `filebeat modules enable zeek && filebeat -e setup`

### Error
```
2020-09-01T12:42:48.781-0500	ERROR	fileset/setup.go:81	Error loading pipeline: 1 error: Error loading pipeline for fileset zeek/x509: couldn't load pipeline: couldn't load json. Error: 400 Bad Request: {"error":{"root_cause [{"type":"parse_exception","reason":"processor [set] doesn't support one or more provided configuration parameters [ignore_empty_value]",
...
```

### With fix:
```
2020-09-01T14:04:56.470-0500	DEBUG	[modules] fileset/pipelines.go:277     	in pipeline filebeat-8.0.0-zeek-x509-pipeline replacing unsupported 'ignore_empty_value' with if ctx?.zeek?.x509?.certificate?.subject?.state != null in set processor
```

### CODE REVIEW

File: filebeat/fileset/pipelines.go [hunk 1]
```diff
@@ -232,3 +237,47 @@ func interpretError(initialErr error, body []byte) error {
...
+		if options, ok := processor["set"].(map[string]interface{}); ok {
+			iev, ok := options["ignore_empty_value"].(bool)
+			if !ok || !iev {
+				continue
+			}
+			val, ok := options["value"].(string)
+			if !ok {
+				continue
+			}
+			newIf := strings.ReplaceAll(val, "{", "")
```
Comment (andrewkroh) [inline filebeat/fileset/pipelines.go:272]: How about `TrimLeft` and `TrimRight` for these two?

File: filebeat/fileset/pipelines.go [hunk 2]
```diff
@@ -232,3 +237,47 @@ func interpretError(initialErr error, body []byte) error {
...
+			newIf = strings.ReplaceAll(newIf, "}", "")
+			newIf = strings.TrimSpace(newIf)
+			newIf = strings.ReplaceAll(newIf, ".", "?.")
+			newIf = "ctx?." + newIf + " != null"
+			logp.Debug("modules", "in pipeline %s replacing unsupported 'ignore_empty_value' with if %s in set processor", pipelineID, newIf)
```
Comment (andrewkroh) [inline filebeat/fileset/pipelines.go:277]:
```suggestion
			logp.Debug("modules", "In pipeline %s replacing unsupported 'ignore_empty_value' with if %q in set processor", pipelineID, newIf)
```

File: filebeat/fileset/pipelines_test.go [hunk 3]
```diff
@@ -215,3 +215,106 @@ func TestSetEcsProcessors(t *testing.T) {
...
+		{
+			name:      "ES < 7.9.0",
+			esVersion: common.MustNewVersion("7.8.0"),
+			content: map[string]interface{}{
+				"processors": []interface{}{
+					map[string]interface{}{
+						"set": map[string]interface{}{
+							"field":              "rule.name",
+							"value":              "{{panw.panos.ruleset}}",
+							"ignore_empty_value": true,
```
Comment (andrewkroh) [inline filebeat/fileset/pipelines_test.go:236]: How about a test case for when the pipeline already has an `if` and it has `ignore_empty_value`. Perhaps it should just drop the `ignore_empty_value` in that case?

File: CHANGELOG.next.asciidoc [hunk 4]
```diff
@@ -256,6 +256,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Fix long registry migration times. {pull}20717[20717] {issue}20705[20705]
 - Fix event types and categories in auditd module to comply with ECS {pull}20652[20652]
 - Update documentation in the azure module filebeat. {pull}20815[20815]
+- provide backwards compatibility for set processor and elasticsearch less than 7.9.0 {pull}20908[20908]
```
Comment (andrewkroh) [inline CHANGELOG.next.asciidoc:259]:
```suggestion
- Provide backwards compatibility for the `set` processor when Elasticsearch is less than 7.9.0. {pull}20908[20908]
```

**Review by andrewkroh — APPROVED** [review]
LGTM. Can you try testing the modules locally against 7.8 if you haven't already.

I think if you `export TESTING_ENVIRONMENT=latest` and roll back to 7.8 in https://github.com/elastic/beats/blob/master/testing/environments/latest.yml#L6-L29 while running the module tests it will use 7.8.

**Comment by leehinman** [comment]
> LGTM. Can you try testing the modules locally against 7.8 if you haven't already.
> 
> I think if you `export TESTING_ENVIRONMENT=latest` and roll back to 7.8 in https://github.com/elastic/beats/blob/master/testing/environments/latest.yml#L6-L29 while running the module tests it will use 7.8.

All tests passed with TESTING_ENVIRONMENT=latest & setting version to 7.8.0 in latest.yml

Also tried loading the zeek module against 7.8.1 & 7.9.0 clusters:


## 7.8.1 cluster

### setup output

```
2020-09-03T10:00:54.917-0500	DEBUG	[modules]	fileset/pipelines.go:270	In pipeline "filebeat-8.0.0-zeek-x509-pipeline" removing unsupported 'ignore_empty_value' in set processor
2020-09-03T10:00:54.917-0500	DEBUG	[modules]	fileset/pipelines.go:288	In pipeline "filebeat-8.0.0-zeek-x509-pipeline" adding if ctx?.zeek?.x509?.certificate?.subject?.common_name != null to replace 'ignore_empty_value' in set processor
2020-09-03T10:00:55.003-0500	INFO	fileset/pipelines.go:139	Elasticsearch pipeline with ID 'filebeat-8.0.0-zeek-x509-pipeline' loaded
```

### Resulting set processor

```
{
        "set" : {
          "value" : "{{zeek.x509.certificate.subject.common_name}}",
          "if" : "ctx?.zeek?.x509?.certificate?.subject?.common_name != null",
          "field" : "file.x509.subject.common_name"
        }
},
```

## 7.9.0 cluster

### setup output

```
2020-09-03T10:21:33.594-0500	INFO	fileset/pipelines.go:139	Elasticsearch pipeline with ID 'filebeat-8.0.0-zeek-x509-pipeline' loaded
```

### Resulting set processor

```
{
        "set" : {
          "ignore_empty_value" : true,
          "field" : "file.x509.subject.common_name",
          "value" : "{{zeek.x509.certificate.subject.common_name}}"
        }
},
```

**Pull request merged**